documentation/projects/proposals/additional_search_views/20230424-project_proposal_additional_search_views.md

```diff
@@ -0,0 +1,36 @@
+# 2023-04-24 Project Proposal
+
+## Reviewers
+
+- [ ] @obulat
+- [ ] @krysal
+
+## Project summary
+
+The Openverse team has decided to enhance the browsing experience by
+incorporating new content views for providers, tags, and creators. This feature
+will allow users to share collections detached of individual searches.
+
+## Goals
+
+Search Experience
+
+## Requirements
+
+## Success
+
+The project will be considered successful when each tag, provider, and creator
+has a standalone and shareable page
+
+## Participants and stakeholders
+
+- Design: @panchovm
+- Implementation:
+
+## Infrastructure
+
+## Accessibility
+
```

Comment:
Do you think we need to add details in the Requirements, Infrastructure, Accessibility, and Required implementation plans sections?

Reply:
Accessibility: we should mention how page structure and headings on the pages affect it.

Implementation plans: we need the design issue (would 1 be enough, or do you plan to have an issue per search view, @panchovm ?) as an implementation plan. I am not sure that we need an IP for the code. Maybe a single issue is enough? I don't think we need any changes in the API, do we?

Reply:
One design issue introducing mockups for each case is suficient. Regarding changes in the API, I would defer to @krysal or ping maintainers folks in case we need more visions.

Reply:
Yeah, I don't think we need changes on the API side since it already provides ways for searching by providers, tags, and creators. The frontend will only need to assemble the url with required query parameters.

Reply:
> I am not sure that we need an IP for the code. Maybe a single issue is enough? I don't think we need any changes in the API, do we?

I think we should have implementation plan(s) for the frontend changes. If the changes are simple then the plan will be small and easy to review and won't add more than a week and a half to the project (maybe less if the process goes faster for a very small IP, as is allowed). We should not skip implementation planning for new views or wholly new designs.

I would suggest two requested issues:

1. The design issue
2. One implementation plan at minimum, to be split depending on the requirements surfaced during design review

> Yeah, I don't think we need changes on the API side since it already provides ways for searching by providers, tags, and creators. The frontend will only need to assemble the url with required query parameters

This is sort of true, but the method for querying is different. When you add individual field queries to the search request, the search term is ignored ([described in detail in our documentation](https://docs.openverse.org/api/reference/search_algorithm.html#individual-field-querying)).

If the design changes allow for doing regular searches against the subset of data, then we would need API changes to accommodate that for sure and it should have its own implementation plan (a third, counting the two I mentioned before).

Anyway, all of that to say, I would make the final list of requested implementation plans dependent on completed designs unless this project proposal is significantly expanded to describe the specific requirements and expected behaviour of the new views. Right now the described requirements are so broad that it's impossible to know what the outcome of the designs would include. This is dangerous because we do not have any guiding light to understand how to cut scope.

Reply:
> If the design changes allow for doing regular searches against the subset of data, then we would need API changes to accommodate that for sure and it should have its own implementation plan (a third, counting the two I mentioned before).

This is a great point to clarify in the proposal! My assumption (based on the what we talked about during the 2023 planning) was that the first iteration that we talk about in **this** project would only include a page with the items that are from the chosen provider, creator or have the selected tag. Roughly, the page would have a large title (tag/provider/creator) and a grid of results for the relevant tag/provider/creator.

I was assuming that it is not possible to do a search within this page. So, you only list all the items with the selected tag, and not search for "parrot" within the images tagged as "parrot".

But I realized after reading the comments that this proposal does not contain information that would support or disprove my assumptions. So, similar to other reviewers, my request is to clarify the scope: is searching within the selected tag/provider/creator supported or not? How are the items sorted (as they come from Elasticsearch, assuming by their popularity score, or by the date when the items were indexed by Openverse)?

Reply:
> This is sort of true, but the method for querying is different. When you add individual field queries to the search request, the search term is ignored ([described in detail in our documentation](https://docs.openverse.org/api/reference/search_algorithm.html#individual-field-querying)).

I just realized that my assumption about the way these views would work was wrong.

I was assuming that it would be a view that does _filtering_ rather than searching. So, if you click on a "cat" tag, you would get all the images with the exact "cat" tag. However, the `tags` parameter in the API *searches* through tags, so for `tags=bot` you could get the results for tags like `robot` or even `botulism`: https://api.openverse.engineering/v1/images/?tags=bot. This is definitely *not* what I would expect when I click on a tag called "bot"!

The same is true for creator (it would return any results for creator whose name _contains_ the search term, not the exact creator), which, again, is not what I would expect as a user.

If everyone agrees with me on the idea that we should return _exact_ matches for tags and creator, I think we *would* need API changes. We would need to add a `filter` similar to `source` or `licenses` that would return exact matches and not search within the field text.

Even the `creator` search suggested in https://github.com/WordPress/openverse/issues/630 would search for text within the creator's name and not return the items from the selected creator. Here's a sample of what you see if you search for creator "K": https://api.openverse.engineering/v1/images/?source=flickr&creator=k

Reply:
Thank you all for your thoughts.

It seems that the designs are mandatory for continuing with the Project Proposal and the Implementation Plan. I am fine with holding this document until having a level of approval for the visual changes. In that vein, the scope of this project is framed within two main user stories:

1. As an openverse user, I want to see all results associated with a single tag.
2. As an openverse user, I want to see all results coming from a single provider.
3. As an openverse user, I want to see all results created by a single author.

That being said, and recalling a preliminary approach we had when suggesting this project, this search view does not allow any search. It _might_ allow sorting by different criteria and display content per content type or in the "all content" grid, but that is a discussion we should have in the design ticket.

At the same time, I agree with @obulat that user expectations would mean exact matches for both tags and authors.

Reply:
If we want to implement exact matches, we would also need an implementation plan for the API changes. Or a section in the Implementation Plan that deals with the API changes in addition to the frontend changes. We would need to add `filter`s for tag and creator. Do you agree, @sarayourfriend ?

As for the provider, we might be able to use the `source` filter. We would need to confirm the if the distinction between `source` and `provider` complicates the things too much. For example, if you decide to click on a provider of a Flickr image from NASA, would you be taken to a Flickr page or a NASA page? Both can be retrieved from the API using `source=flickr` or `source=NASA`.

Reply:
Can you clarify what you mean by "add filters for tag and creator"? Are you talking about the ES boolean query "filter" feature or something else?

Exact match queries should be possible using quotations because we escape quotes, but there's virtually no guarantee that that works as expected: https://github.com/WordPress/openverse/blob/697f62f01a32cb7fcf2f4a7627650a113cba40da/api/catalog/api/controllers/search_controller.py#L373

Indeed, I get the same set of results (at least at a glance) when I query tags for `"bot"` and `bot` (with and without quotes).

> It seems that the designs are mandatory for continuing with the Project Proposal and the Implementation Plan

I think we should define the scope of expected behaviour before design work starts in earnest, mostly to give you, Francisco, and anyone reviewing the designs guidance for preventing scope creep while ensuring they are feature complete.